libraries-3/src/test/java/com/baeldung/jnats/NatsClientLiveTest.java

```diff
@@ -1,115 +1,336 @@
+            .reportNoResponders()
+            .build();
+        return Nats.connect(options);
+    }
 
-        client.publishMessage(message.getReplyTo(), message.getSubject(), "hello back");
+    static class CustomErrorListener implements ErrorListener {
+        @Override
+        public void errorOccurred(Connection conn, String error) {
+            System.err.println("Error Occurred: " + error);
+        }
 
-        message = barSubscription.nextMessage(200);
-        assertNotNull("No message!", message);
-        assertEquals("hello back", new String(message.getData()));
+        @Override
+        public void exceptionOccurred(Connection conn, Exception exp) {
+            System.err.println("Exception Occurred: " + exp);
+        }
     }
 
+//    public static void main(String[] args) throws Exception {
+//        Connection conn = createConnection("nats://localhost:4222");
+//        Thread.sleep(60000);
+//    }
 
-    private NatsClient connectClient() {
-        return new NatsClient();
+    @Test
+    public void givenNoSubscribers_whenPublishingMessages_thenItCannotBeDeterminedIfMessageIsReceived() {
+        try (Connection natsConnection = createConnection()) {
+            natsConnection.publish("mySubject", "myData".getBytes());
+        }
+        catch (Exception e) {
+            fail("There should not be an exception");
+        }
     }
 
     @Test
-    public void whenWildCardSubscription_andMatchTopic_MessageReceived() throws Exception {
+    public void givenNoSubscribersAndReportNoResponders_whenRequestReply_thenReceiveNoResponders() {
+        try (Connection natsConnection = createConnectionWithReportNoResponders()) {
+            natsConnection.publish("noOneIsSubscribed", null);
+            CompletableFuture<Message> f = natsConnection.request("noOneIsSubscribed", null);
+            f.get(WAIT_AT_MOST_MILLIS, TimeUnit.MILLISECONDS);
+            fail("There should be an exception");
+        }
+        catch (ExecutionException e) {
+            Throwable cause = e.getCause();
+            assertNotNull(cause);
+            assertInstanceOf(JetStreamStatusException.class, cause);
+            assertTrue(cause.getMessage().contains("503 No Responders Available For Request"));
+        } catch (InterruptedException | TimeoutException | IOException e) {
+            fail("There should not be one of these exceptions");
+        }
+    }
 
-        NatsClient client = connectClient();
+    @Test
+    public void whenSubscribingSynchronouslyWithoutQueueGroups_thenEachSubscriptionShouldReceiveEachMessage() throws Exception {
+        try (Connection natsConnection = createConnection()) {
+            Subscription subscription1 = natsConnection.subscribe("mySubject");
+            Subscription subscription2 = natsConnection.subscribe("mySubject");
 
-        SyncSubscription fooSubscription = client.subscribeSync("foo.*");
+            natsConnection.publish("mySubject", "data".getBytes());
```

Comment:
We should really be passing a character set like StandardCharSets.UTF_8 whenever we do String.getBytes and new String, to ensure the code is portable

(similar in a few other places throughout the test)

Reply:
fixed all instances with
```
getBytes(StandardCharsets.UTF_8)
```

Reply:
And the same for new String

Reply:
made conversions helper functions for all string to bytes and bytes to string